operators/pkg/instance-controller/logic.go

```diff
@@ -43,76 +36,151 @@ func (r *InstanceReconciler) CreateVMEnvironment(instance *crownlabsv1alpha2.Ins
+	})
+	if err != nil {
 		r.setInstanceStatus(ctx, "Could not create secret "+secret.Name+" in namespace "+secret.Namespace, "Warning", "SecretNotCreated", instance, "", "")
 	} else {
 		r.setInstanceStatus(ctx, "Secret "+secret.Name+" correctly created in namespace "+secret.Namespace, "Normal", "SecretCreated", instance, "", "")
 	}
 
 	// create Service to expose the vm
-	service := instance_creation.ForgeService(name, namespace, globalOwnerReference)
-	if err = instance_creation.CreateOrUpdate(ctx, r.Client, service); err != nil {
+	service := instance_creation.ForgeService(name, namespace)
+	_, err = ctrl.CreateOrUpdate(ctx, r.Client, &service, func() error {
+		return ctrl.SetControllerReference(instance, &service, r.Scheme)
+	})
+	if err != nil {
 		r.setInstanceStatus(ctx, "Could not create service "+service.Name+" in namespace "+service.Namespace, "Warning", "ServiceNotCreated", instance, "", "")
 		return err
 	}
 	r.setInstanceStatus(ctx, "Service "+service.Name+" correctly created in namespace "+service.Namespace, "Normal", "ServiceCreated", instance, "", "")
 
 	urlUUID := uuid.New().String()
 	// create Ingress to manage the service
-	ingress := instance_creation.ForgeIngress(name, namespace, &service, urlUUID, r.WebsiteBaseURL, globalOwnerReference)
-	if err = instance_creation.CreateOrUpdate(ctx, r.Client, ingress); err != nil {
+	ingress := instance_creation.ForgeIngress(name, namespace, &service, urlUUID, r.WebsiteBaseURL)
+	_, err = ctrl.CreateOrUpdate(ctx, r.Client, &ingress, func() error {
+		return ctrl.SetControllerReference(instance, &ingress, r.Scheme)
+	})
+	if err != nil {
 		r.setInstanceStatus(ctx, "Could not create ingress "+ingress.Name+" in namespace "+ingress.Namespace, "Warning", "IngressNotCreated", instance, "", "")
 		return err
 	}
 	r.setInstanceStatus(ctx, "Ingress "+ingress.Name+" correctly created in namespace "+ingress.Namespace, "Normal", "IngressCreated", instance, "", "")
 
-	if err = r.createOAUTHlogic(name, instance, namespace, globalOwnerReference, urlUUID); err != nil {
+	if err = r.createOAUTHlogic(name, instance, namespace, urlUUID); err != nil {
 		return err
 	}
 
-	// create VirtualMachineInstance
-	vmi, err := instance_creation.CreateVirtualMachineInstance(name, namespace, environment, instance.Name, secret.Name, globalOwnerReference)
-	if err != nil {
-		return err
+	// create vm
+	var vmname string
+	if environment.Persistent {
+		vm := v1.VirtualMachine{ObjectMeta: metav1.ObjectMeta{Name: name, Namespace: namespace}}
+		_, err = ctrl.CreateOrUpdate(ctx, r.Client, &vm, func() error {
+			instance_creation.UpdateVirtualMachineSpec(environment, name, &vm, instance)
+			instance_creation.UpdateVMLabels(&vm, instance.Name, environment)
+			return ctrl.SetControllerReference(instance, &vm, r.Scheme)
+		})
+		vmname = vm.Name
+		vmi = &v1.VirtualMachineInstance{ObjectMeta: metav1.ObjectMeta{
+			Name:      vmname,
+			Namespace: namespace,
+		}}
```

Comment:
This could be moved before the if, as present in both branches. The `vmname` variable is probably redundant